gramps_webapi/api/resources/filters.py

```diff
@@ -0,0 +1,298 @@
+            add_filter = True
+            if args.get("filter") and args["filter"] != filter_class.get_name():
+                add_filter = False
+            if add_filter:
+                filter_list.append(
+                    {
+                        "comment": filter_class.get_comment(),
+                        "function": filter_class.get_logical_op(),
+                        "invert": filter_class.invert,
+                        "name": filter_class.get_name(),
+                        "rules": [
+                            {
+                                "name": filter_rule.__class__.__name__,
+                                "regex": filter_rule.use_regex,
+                                "values": filter_rule.values(),
+                            }
+                            for filter_rule in filter_class.get_rules()
+                        ],
+                    }
+                )
+        if args.get("filter") and not args.get("rule"):
+            return self.response(200, {"filters": filter_list})
+        return self.response(200, {"filters": filter_list, "rules": rule_list})
+
+    @use_args(FilterSchema(), location="json")
+    def post(self, args: Dict, namespace: str) -> Response:
+        """Create a custom filter."""
+        try:
+            namespace = GRAMPS_NAMESPACES[namespace]
+        except KeyError:
+            abort(404)
+
+        new_filter = build_filter(args, namespace, keys="long")
+        filters.reload_custom_filters()
+        for filter_rule in filters.CustomFilters.get_filters(namespace):
+            if new_filter.get_name() == filter_rule.get_name():
+                abort(422)
+        filters.CustomFilters.add(namespace, new_filter)
+        filters.CustomFilters.save()
+        return self.response(201, {"message": "Added filter: " + new_filter.get_name()})
+
+    @use_args(FilterSchema(), location="json")
+    def put(self, args: Dict, namespace: str) -> Response:
+        """Update a custom filter."""
+        try:
+            namespace = GRAMPS_NAMESPACES[namespace]
+        except KeyError:
+            abort(404)
+
+        new_filter = build_filter(args, namespace, keys="long")
+        filters.reload_custom_filters()
+        for filter_rule in filters.CustomFilters.get_filters(namespace):
+            if new_filter.get_name() == filter_rule.get_name():
+                filters.CustomFilters.get_filters(namespace).remove(filter_rule)
+                filters.CustomFilters.add(namespace, new_filter)
+                filters.CustomFilters.save()
+                return self.response(
+                    200, {"message": "Updated filter: " + new_filter.get_name()}
+                )
+        return abort(404)
+
+    @use_args(
+        {
+            "name": fields.Str(required=True),
+            "force": fields.Boolean(required=False, missing=False),
+        },
+        location="json",
+    )
+    def delete(self, args: Dict, namespace: str) -> Response:
```

Comment:
It's great that you're using type hints everywhere, it makes the code much more easy to understand.

For dicts, it's also helpful to indicate the type of the keys (and possibly values); this also catches possible bugs early in my experience. In this case, problably `Dict[str, str]`.

Reply:
Yes, I am trying to get in the practice of doing so, they are helpful. Good idea on the Dict hints, did not occur to me, I will try to do that as well.